source/common/router/router.cc

```diff
@@ -778,5 +763,37 @@ ProdFilter::createRetryState(const RetryPolicy& policy, Http::HeaderMap& request
                                 priority);
 }
 
+void Filter::UpstreamRequest::setRequestEncoder(Http::StreamEncoder& request_encoder) {
+  request_encoder_ = &request_encoder;
+  // Now that there is an encoder, have the connection manager inform the manager when the
+  // downstream buffers are overrun.  This may result in immediate watermark callbacks referencing
+  // the encoder.
+  parent_.callbacks_->addDownstreamWatermarkCallbacks(downstream_watermark_manager_);
+}
+
+void Filter::UpstreamRequest::clearRequestEncoder() {
+  // Before clearing the encoder, unsubscribe from callbacks.  This may result
+  // in some clean up calls to onBelowWriteBufferLowWatermark which reference
```

Comment:
For HTTP/1 we reuse connections IFF the the response was complete (otherwise we destroy the connection). (Standard HTTP/1 conn pool semantics). Since you do the clear/cleanup inline with either reset or end_stream, I think this is probably safe if you have the right guards during stack unwind, but honestly I'm not totally sure without looking at it more. It may end up being simpler to clean up disable inside the codec directly or in the codec client (whatever we end up doing please try to have as much symmetry as possible between HTTP/1 and HTTP/2 and upstream and downstream per some of my comments in the other PRs).

Either method is fine with me, I would just recommend trying to get tests that cause crashes (or assertion failures) with the stacks we have already seen in production.